### PR TITLE
AV-65723 #comment Add method to do get_all operaion

### PR DESCRIPTION
**Pull request by sabandi**

### CODE REVIEW

File: python/avi/sdk/avi_api.py [hunk 1]
```diff
@@ -931,6 +931,35 @@ def get_obj_uuid(self, obj):
...
+
+        page = 2 #  Initialized to 2 for new page requests
+        while resp.get('next', None) is not None:
+            params['page'] = page
+            resp = self.get(obj_type, tenant, tenant_uuid, timeout=timeout,
+                            params=params, api_version=api_version, **kwargs).json()
+            all_objs += resp['results']
+            page += 1
+
+        for obj in all_objs:
```
Comment (sabandi) [inline python/avi/sdk/avi_api.py]: @grastogi23  I am not sure what benefit we get by using this generator model as we anyways have all the objects in memory.

We could do a simple `return all_objs` .
Please let me know your thoughts.

Comment (grastogi23) [reply python/avi/sdk/avi_api.py]: we should not keep all objects in the memory... The semantics is to provide an iterator not fetch all objects.

File: python/avi/sdk/avi_api.py [hunk 2]
```diff
@@ -931,6 +931,35 @@ def get_obj_uuid(self, obj):
         else:
             return None
 
+    def get_all_objects_of_type(self, obj_type, tenant='', tenant_uuid='',
+                                timeout=None, params=None, api_version=None,
```
Comment (grastogi23) [inline python/avi/sdk/avi_api.py]: should document the parameters. Also provide an example use of this api.

let us call it get_objects_iter so it is explicit that it is not return all objects but an iterator to iterate over objects.

**Review by grastogi23 — CHANGES_REQUESTED** [review]
please see inline comments.

File: python/avi/sdk/avi_api.py [hunk 3]
```diff
@@ -931,6 +931,35 @@ def get_obj_uuid(self, obj):
...
+                                **kwargs):
+        """
+        Helper method to get all objects of specific type from controller.
+        """
+        all_objs = []
+        if params is None:
+            params={}
+        params['page_size'] = 200 # Max limt on Controller
```
Comment (grastogi23) [inline python/avi/sdk/avi_api.py]: let us take the page_size from the params as well. But, we can check for max of 200

File: python/avi/sdk/avi_api.py [hunk 4]
```diff
@@ -931,6 +931,35 @@ def get_obj_uuid(self, obj):
...
+            all_objs += resp['results']
+        else:
+            # For apis returning single object eg. api/cluster
+            all_objs += [resp]
+
+        page = 2 #  Initialized to 2 for new page requests
+        while resp.get('next', None) is not None:
+            params['page'] = page
+            resp = self.get(obj_type, tenant, tenant_uuid, timeout=timeout,
+                            params=params, api_version=api_version, **kwargs).json()
```
Comment (grastogi23) [inline python/avi/sdk/avi_api.py]: what happens in case of error. It should raise exception?

File: python/avi/sdk/avi_api.py [hunk 5]
```diff
@@ -931,6 +931,83 @@ def get_obj_uuid(self, obj):
...
+        on top of the pagination functionality provided by Avi Controller.
+
+        :param *args: Accepts all args accepted by get method of class
+            ApiSession.
+        :params **kwargs: Accepts all kwargs accepted by get method of class
+            ApiSession.
+
+        Yields batch of objects in the range of 0 to page_size passed in kwargs.
+        """
+        resp = self.get(*args, **kwargs).json()
```
Comment (sabandi) [inline python/avi/sdk/avi_api.py:946]: .json() method in avi_api.py checks and raises various exceptions. Hence left the exception handling as it is already taken care of in json() method @ https://github.com/avinetworks/sdk/blob/bba674f45ce80374f2491a24b2137050cf86f70a/python/avi/sdk/avi_api.py#L89

**Review by grastogi23 — APPROVED** [review]

**Review by chaitanyaavi — APPROVED** [review]

**Review by grastogi23 — APPROVED** [review]

**Pull request merged**